### ISSUE TITLE
How production-ready is purerl now?

### ISSUE THREAD

**Issue by drathier** [purerl/purerl#13]
It looks from outside like it's pretty much as production ready as purescript itself is. What are the main purerl-specific rough edges right now? Is anyone using it in production?

**Comment by adrianroe** [comment]
We (https://id3as.co.uk) have purerl live in production with multiple customers and are utterly delighted with the results.  That's not to say there aren't rough edges, but (so long as you know Erlang as well) these can normally be easily worked around with FFI.  We are steadily implementing more and more ergonomic interfaces on top of OTP / Cowboy etc (take a look at the way e.g. https://github.com/id3as/purescript-erl-pinto is evolving) and there are likely to be more breaking changes as those interfaces evolve.  Every now and then we also need to implement wrappers on existing Erlang libraries - the ecosystem is not mature yet - so that probably counts as a rough edge too.

You also have to be careful with performance in certain cases - it is easy to have performance issues due to heap churn in very hot code paths - but that's true of Erlang as well.  For 99% of code we can (and do) completely ignore performance.  There is some really cool work around memoization happening that looks like it might improve performance even for some of the edge cases.  For code that is truly performance sensitive we typically reach for Rust anyway.

In short - yes we do use it in production and it would take something seismic to happen to make us look elsewhere - it's a fantastic environment to build large scale, fault tolerant applications that you can fearlessly refactor, knowing the compiler has your back.

**Comment by adrianroe** [comment]
Purerl (the compiler) has been remarkably bug-free.  Off the top of my head we've come across one bug so far (something rather esoteric to do with variable scoping) but it didn't impact any production systems.

**Comment by robashton** [comment]
From my own perspective  (as author of Stetson/Pinto) and subsequent user of those tools to write rather a lot of Purerl one of  our newer systems, the experience has been an interesting one as we've had to come up with a few patterns to work with Erlang in ways that are different than how Purescript would have been written for Javascript.

It has been a while since I encountered code that needed any of those decision points and while Adrian points out that breaking changes are forthcoming in Pinto/Stetson that's largely as a result of our own learnings on better ways of writing Purescript APIs and not as a result of anything  being broken - "wouldn't it be nicer if we did it this way".

I am increasingly confident of recommending Purerl to people wishing to write new code on top of the BEAM, I'd highly recommend using a nix environment to do that in (as the main pain point for me is "getting started" and that somewhat solves that - https://purerl-cookbook.readthedocs.io/en/latest/devenv/nix.html ). 

The rest is just "Oh, FFI needs writing for such-and-such" and deciding whether that is worth turning into a library for others or if you only need a single function and actually let's not bother. Hopefully over time that'll become less of a thing as more FFI gets written and published for others to use.

**Comment by drathier** [comment]
Do you use purerl inside modules for business logic and otp for message passing and fault tolerance, or does the purescript deal with otp and message passing as well?

**Comment by robashton** [comment]
Purerl simply compiles into Erlang that then leverages what OTP is providing.  In the case of supervision trees/etc - that functionality is surfaced in  Erl.Pinto and in the case of simple message passing, Erl.Process.

Hence circling back to "production readiness" we can be reasonably confident that code written in Purerl is going to do what you think it does, because it's not re-implementing any of this stuff from scratch.

**Comment by drathier** [comment]
Sorry, what I meant was, do you generally interact with OTP and message passing from purescript code, or just use purescript for business logic inside modules and ffi/erlang/elixir to handle message passing and such? It sounds like you use purescript even across module boundaries. https://purerl-cookbook.readthedocs.io/en/latest/otp/dynamic_children.html 

If you do cross module boundaries with purescript code, is there a good way to deal with type safety across modules when hot code reloading?

**Comment by nwolverson** [comment]
There's no solution just now for type-safe code hot-loading with purerl - it would certainly be cool to leverage types to determine (as far as possible) whether it is safe to do module replacement, but that's not a thing that exists just now.

Having said that, with a bit of care things should work out fine with code replacement in situations where it's possible - eg replacing a Pinto gen server. We're not doing whole-program compilation, a PureScript module is 1:1 with an erlang module - I'd just be careful with non-local changes (module dependencies, or typeclass stuff).

**Comment by drathier** [comment]
Okey, it sounds to me like purerl is production ready :) Thanks for all the input everyone!

**Comment by nwolverson** [comment]
I'm going to close this issue as it seems like there is some resolution to the original question (though I'd stress that the concept "production ready" is ultimately a personal or context-dependant cost/benefit decision). Not to say further commentary is not invited.

For general discussion/assistance/misc questions the #purerl channel on FP slack is the place to go - in fact I'm just about to add that to the README here
